### PR TITLE
Implemented TailCall ELT hook for arm32 Linux

### PR DESCRIPTION
**Pull request by rartemev**
Issue #13993
Implementation is the same as LEAVE hook except stub using in it. The windows implementation is also similar to LEAVE and already exists.

Unfortunately tailcall optimization is not implemented in JIT for arm32 (see src/jit/target.h: 1194/1195) the real goal of this PR to add this hook by request from diagnostic team.

### CODE REVIEW

**Comment by rartemev** [comment]
@dotnet/jit-contrib PTAL

**Comment by sywhang** [comment]
These hooks have not yet been tested against the profiler API tests due to the missing tailcall optimization that Roman mentioned. but in case we happen to add the tailcall optimization for arm32 later, we're adding this hook in first. @lt72 do you have any comments?

**Comment by BruceForstall** [comment]
@janvorli PTAL

File: src/vm/arm/asmhelpers.S [hunk 1]
```diff
@@ -741,6 +741,49 @@ NESTED_ENTRY ProfileLeaveNaked, _TEXT, NoHandler
...
+//
+NESTED_ENTRY ProfileTailcallNaked, _TEXT, NoHandler
+    PROLOG_PUSH "{r1, r2, r4, r5, r7, r11, lr}"
+    PROLOG_STACK_SAVE_OFFSET r7, #16
+
+    // fields of PLATFORM_SPECIFIC_DATA, in reverse order
+
+    // UINT32      r0;         // Keep r0 & r1 contiguous to make returning 64-bit results easier
+    // UINT32      r1;
+    // void       *R11;
```
Comment (janvorli) [inline src/vm/arm/asmhelpers.S]: A tiny nit - R11 -> r11

File: src/vm/arm/asmhelpers.S [hunk 2]
```diff
@@ -741,6 +741,49 @@ NESTED_ENTRY ProfileLeaveNaked, _TEXT, NoHandler
...
+    add         r5, r11, #8
+    push        { /* profiledSp */ r5 }
+    add         r5, sp, #40
+    push        { /* probeSp    */ r5 }
+    push        { /* functionId */ r0 }
+    vpush.64    { d0 - d7 }
+    push        { lr }
+    push        { r11 }
+    push        { r1 }
+    push        { r2 }
```
Comment (janvorli) [inline src/vm/arm/asmhelpers.S]: The comment above says `r0` is here, but the code pushes `r2`

File: src/vm/arm/asmhelpers.S [hunk 3]
```diff
@@ -741,6 +741,49 @@ NESTED_ENTRY ProfileLeaveNaked, _TEXT, NoHandler
...
 NESTED_END ProfileLeaveNaked, _TEXT
 
+//
+// EXTERN_C void ProfileTailcallNaked(FunctionIDOrClientID functionIDOrClientID);
+//
+NESTED_ENTRY ProfileTailcallNaked, _TEXT, NoHandler
+    PROLOG_PUSH "{r1, r2, r4, r5, r7, r11, lr}"
+    PROLOG_STACK_SAVE_OFFSET r7, #16
+
+    // fields of PLATFORM_SPECIFIC_DATA, in reverse order
```
Comment (janvorli) [inline src/vm/arm/asmhelpers.S]: The struct name is  `PROFILE_PLATFORM_SPECIFIC_DATA`, not `PLATFORM_SPECIFIC_DATA`. And the comment "in reverse order" is misleading. The fields are in the order that you've listed them below. I guess you've meant that they are pushed in the reverse order by that comment, right?

Comment (rartemev) [reply src/vm/arm/asmhelpers.S]: Fixed

File: src/vm/arm/asmhelpers.S (hunk 1 above)
Comment (rartemev) [reply src/vm/arm/asmhelpers.S]: Fixed

File: src/vm/arm/asmhelpers.S (hunk 2 above)
Comment (rartemev) [reply src/vm/arm/asmhelpers.S]: Fixed

**Comment by lt72** [comment]
Great to have the hook available for testing!

**Comment by rartemev** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by janvorli** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test please

**Comment by janvorli** [comment]
The Windows tests failures are unrelated

File: src/vm/arm/asmhelpers.S (hunk 2 above)
Comment (kvochko) [reply src/vm/arm/asmhelpers.S]: I believe this should be r2 for the LEAVE hook. The profiled function's return value is passed in that register:
https://github.com/dotnet/coreclr/blob/5427b66d02cecd6990229744e4c9fb0dce9b7d3c/src/jit/codegencommon.cpp#L8175-L8176

**Pull request merged**